{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tb TF KO expression data wrangling\n",
    "\n",
    "Created by Emanuel Flores and Adrian Jinich. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T05:46:49.895332Z",
     "start_time": "2020-02-02T05:46:49.881846Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "\n",
    "# black magic for code style \n",
    "%load_ext blackcellmagic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Welcome! In this notebook we're going to do some simple data selection in order to make a dataset that we can work with later on. The data we'll be working with is fold-change (FC) data using single transcription factor (TF) overexpression. What this means is that they measured gene expression of all the transcriptome in normal conditions and contrasted that with the conditions of overexpressing each TF in the network, and they computed the $\\mathrm{log} \\left(\\frac{g_{i, KO}}{g_i} \\right)$, where $g_i$ is the relative level of expression of gene $i$. We're going to merge this dataset with annotation from the oxidorreductases data. \n",
    "\n",
    "First, we're going to load the fold-change data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:41:30.302341Z",
     "start_time": "2020-02-02T01:40:54.220300Z"
    }
   },
   "outputs": [],
   "source": [
    "# Using pandas to load data.\n",
    "fc_xl = pd.read_excel('~/Downloads/fold_changes_SI_Table_2.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:41:37.402981Z",
     "start_time": "2020-02-02T01:41:37.332152Z"
    }
   },
   "outputs": [],
   "source": [
    "# take a look\n",
    "fc_xl.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T02:07:04.981442Z",
     "start_time": "2020-02-02T02:07:04.971757Z"
    }
   },
   "outputs": [],
   "source": [
    "fc_xl.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that there are some weird entries with NaN values. Let's filter only the FC dat.a "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter only fold-change values "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because the dataset contains p-values and ChIP-seq binding information,we'll filter the dataset to get only FC data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:41:51.098061Z",
     "start_time": "2020-02-02T01:41:51.090802Z"
    }
   },
   "outputs": [],
   "source": [
    "# Getting column names for fold-change data. \n",
    "fc_columns = [col for col in fc_xl.columns.to_list() if 'FC' in col]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We know that there are 206 TFs, let's make sure we got all of the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T02:17:06.285489Z",
     "start_time": "2020-02-02T02:17:06.275417Z"
    }
   },
   "outputs": [],
   "source": [
    "# Make sure we got 206 columns \n",
    "len(fc_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:41:52.659309Z",
     "start_time": "2020-02-02T01:41:52.652975Z"
    }
   },
   "outputs": [],
   "source": [
    "# Filtering annotation data \n",
    "annot = fc_xl.iloc[:, :4]\n",
    "\n",
    "annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:42:15.824234Z",
     "start_time": "2020-02-02T01:42:15.818898Z"
    }
   },
   "outputs": [],
   "source": [
    "# Merging column names with annotation or FC data into a list\n",
    "cols = annot.columns.to_list() + fc_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:42:32.394611Z",
     "start_time": "2020-02-02T01:42:32.380218Z"
    }
   },
   "outputs": [],
   "source": [
    "# Selection of fold-change data using fancy indexing on columns \n",
    "fc_data = fc_xl[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T02:18:02.538911Z",
     "start_time": "2020-02-02T02:18:02.507660Z"
    }
   },
   "outputs": [],
   "source": [
    "# take a look \n",
    "fc_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:42:38.167590Z",
     "start_time": "2020-02-02T01:42:38.159146Z"
    }
   },
   "outputs": [],
   "source": [
    "fc_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nice, now we can save the dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T01:42:40.345360Z",
     "start_time": "2020-02-02T01:42:40.265629Z"
    }
   },
   "outputs": [],
   "source": [
    "#fc_data.to_csv('../data/fold_change_tf_ko.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load redox annotation data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we want to integrate the fold-change dataset with the redox annotation, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:10:04.002581Z",
     "start_time": "2020-02-02T03:10:03.848543Z"
    }
   },
   "outputs": [],
   "source": [
    "redox = pd.read_excel('~/Downloads/redox_uniprot_myco.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:10:04.386074Z",
     "start_time": "2020-02-02T03:10:04.378231Z"
    }
   },
   "outputs": [],
   "source": [
    "redox['redox_enzyme'] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's make an outer join keeping all the data from both dataframes. \n",
    "\n",
    "In order to do that, we'll first have to change the column name of the FC data to Rv_ID, to join on that column as an anchor. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:10:06.141152Z",
     "start_time": "2020-02-02T03:10:06.126862Z"
    }
   },
   "outputs": [],
   "source": [
    "# Change to have the locus name, Rv_ID as an anchor \n",
    "#fc_data.rename(columns = {'Name': 'gene_name', 'ID': 'Rv_ID'}, inplace = True)\n",
    "\n",
    "# Change redox table column names for consistency \n",
    "redox.rename(\n",
    "    columns={\"Gene names\": \"alternative_gene_names\", \"Name\": \"redox_gene_name\", \n",
    "             'Function': 'function_redox_'},\n",
    "    inplace=True,\n",
    ")\n",
    "\n",
    "# Delete organism column because we know we're working with TB \n",
    "del(redox['Organism'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:10:07.992151Z",
     "start_time": "2020-02-02T03:10:07.984874Z"
    }
   },
   "outputs": [],
   "source": [
    "# Confirm the only anchor column is the locus tag \n",
    "common_cols = list(set(redox.columns.tolist()).intersection(fc_data.columns.tolist()))\n",
    "\n",
    "common_cols"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All right, we're ready to do the merge. We'll do a left outer join because we know (after some analysis) that there are some genes in the redox dataset not in the fold change dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:16:37.282967Z",
     "start_time": "2020-02-02T03:16:37.273236Z"
    }
   },
   "outputs": [],
   "source": [
    "# Get genes not in FC dataset\n",
    "set(redox.Rv_ID.values) - set(fc_data.Rv_ID.values).intersection(set(redox.Rv_ID.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:16:40.850943Z",
     "start_time": "2020-02-02T03:16:40.812090Z"
    }
   },
   "outputs": [],
   "source": [
    "df = pd.merge(fc_data, redox, on = 'Rv_ID', how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:16:45.245801Z",
     "start_time": "2020-02-02T03:16:45.237045Z"
    }
   },
   "outputs": [],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we only need to populate some annotation entries with zeros. This will help later for classification and visualization purposes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:17:44.157585Z",
     "start_time": "2020-02-02T03:17:44.142546Z"
    }
   },
   "outputs": [],
   "source": [
    "# Fill the entries of the non-redox annotated proteins with 0s\n",
    "df.redox_enzyme.fillna(0, inplace = True)\n",
    "df.UK_score_4.fillna(0, inplace = True)\n",
    "df.Annotation.fillna('None', inplace = True)\n",
    "df.Annotation_int.fillna(0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-02T03:17:46.045583Z",
     "start_time": "2020-02-02T03:17:44.736443Z"
    }
   },
   "outputs": [],
   "source": [
    "# Save dataset \n",
    "df.to_csv('../data/fold_change_tf_ko_plus_redox_annot.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n",
    "\n",
    "%watermark -v -p pandas"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
