{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "load TnSeq / expression data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_data = '../data/integrated_rnaseq_tnseq_v1.csv'\n",
    "df_data = pd.read_csv(file_data)\n",
    "cols = df_data.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1173"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many columns\n",
    "len(cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rv_ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Geneid/Contrast_id</th>\n",
       "      <th>is_sdr</th>\n",
       "      <th>Function</th>\n",
       "      <th>groups</th>\n",
       "      <th>redox_enzyme</th>\n",
       "      <th>function_redox_</th>\n",
       "      <th>UK_score_4</th>\n",
       "      <th>function_sanger</th>\n",
       "      <th>...</th>\n",
       "      <th>zhang_AA_Rescue_vs_zhang_in_vitro_control_Rescue</th>\n",
       "      <th>zhang_DETA-NO_pH_7.0_vs_zhang_pH_7.0_no_NO_control</th>\n",
       "      <th>zhang_Fe_1.5mM_vs_zhang_Fe_450uM</th>\n",
       "      <th>zhang_Trp_Rescue_vs_zhang_in_vitro_control_Rescue</th>\n",
       "      <th>zhang_Tyloxapol_pH_6.5_vs_zhang_Tyloxapol_pH_4.5</th>\n",
       "      <th>zhang_Tyloxapol_pH_6.5_vs_zhang_pcit_pH_4.5</th>\n",
       "      <th>zhang_mhcii_mouse_d10_vs_zhang_wt_mouse_d10</th>\n",
       "      <th>zhang_mhcii_mouse_d45_vs_zhang_wt_mouse_d45</th>\n",
       "      <th>zhang_wt_mouse_d10_vs_zhang_input_library</th>\n",
       "      <th>zhang_wt_mouse_d45_vs_zhang_input_library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Chromosomal replication initiator protein DnaA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chromosomal replication initiator protein</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 1173 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Rv_ID Gene name  Geneid/Contrast_id  is_sdr  \\\n",
       "0  Rv0001      dnaA                 1.0     0.0   \n",
       "\n",
       "                                         Function  groups  redox_enzyme  \\\n",
       "0  Chromosomal replication initiator protein DnaA     0.0           0.0   \n",
       "\n",
       "  function_redox_  UK_score_4                            function_sanger  \\\n",
       "0             NaN         0.0  chromosomal replication initiator protein   \n",
       "\n",
       "                     ...                      \\\n",
       "0                    ...                       \n",
       "\n",
       "  zhang_AA_Rescue_vs_zhang_in_vitro_control_Rescue  \\\n",
       "0                                              0.0   \n",
       "\n",
       "   zhang_DETA-NO_pH_7.0_vs_zhang_pH_7.0_no_NO_control  \\\n",
       "0                                                0.0    \n",
       "\n",
       "   zhang_Fe_1.5mM_vs_zhang_Fe_450uM  \\\n",
       "0                               0.0   \n",
       "\n",
       "   zhang_Trp_Rescue_vs_zhang_in_vitro_control_Rescue  \\\n",
       "0                                                0.0   \n",
       "\n",
       "   zhang_Tyloxapol_pH_6.5_vs_zhang_Tyloxapol_pH_4.5  \\\n",
       "0                                               0.0   \n",
       "\n",
       "   zhang_Tyloxapol_pH_6.5_vs_zhang_pcit_pH_4.5  \\\n",
       "0                                          0.0   \n",
       "\n",
       "   zhang_mhcii_mouse_d10_vs_zhang_wt_mouse_d10  \\\n",
       "0                                          0.0   \n",
       "\n",
       "   zhang_mhcii_mouse_d45_vs_zhang_wt_mouse_d45  \\\n",
       "0                                          0.0   \n",
       "\n",
       "   zhang_wt_mouse_d10_vs_zhang_input_library  \\\n",
       "0                                        0.0   \n",
       "\n",
       "   zhang_wt_mouse_d45_vs_zhang_input_library  \n",
       "0                                        0.0  \n",
       "\n",
       "[1 rows x 1173 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting all data for a single gene: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_rna_annot = '../data/experiment_annotation_master_colombos_v0.csv'\n",
    "df_rna_annot = pd.read_csv(file_rna_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_data = np.array(df_data.columns[11:].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "inds_rnaseq = [i for i in range(len(cols_data)) if not 'vs' in cols_data[i]]\n",
    "inds_tnseq = [i for i in range(len(cols_data)) if 'vs' in cols_data[i]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_data(rv, df_data, cols_data):\n",
    "    df_rv = df_data[df_data.Rv_ID==rv]\n",
    "    rv_data = df_rv[cols_data].values[0]\n",
    "    return rv_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "rv_1 = 'Rv3400'\n",
    "rv_1_data = get_gene_data(rv_1, df_data, cols_data)\n",
    "\n",
    "rv_2 = 'Rv3410c'\n",
    "rv_2_data = get_gene_data(rv_2, df_data, cols_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"significant\" RNA screens:\n",
    "rv_1_inds_rnaseq_sig = [i for i in inds_rnaseq if np.abs(rv_1_data[i])>=1]\n",
    "rv_2_inds_rnaseq_sig = [i for i in inds_rnaseq if np.abs(rv_2_data[i])>=1]\n",
    "\n",
    "# \"significant\" TnSeq screens:\n",
    "rv_1_inds_tnseq_sig = [i for i in inds_tnseq if np.abs(rv_1_data[i])==1]\n",
    "rv_2_inds_tnseq_sig = [i for i in inds_tnseq if np.abs(rv_2_data[i])==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'345'}"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cols_data[rv_1_inds_rnaseq_sig]).intersection(set(cols_data[rv_2_inds_rnaseq_sig]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'carey_621_vs_mbio_H37Rv', 'carey_631_vs_mbio_H37Rv'}"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cols_data[rv_1_inds_tnseq_sig]).intersection(set(cols_data[rv_2_inds_tnseq_sig]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contrast_name</th>\n",
       "      <th>annotation</th>\n",
       "      <th>Test description</th>\n",
       "      <th>Reference description</th>\n",
       "      <th>Experiment_id</th>\n",
       "      <th>Data source</th>\n",
       "      <th>Platform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>GSM28131.ch1-vs-GSM28131.ch2</td>\n",
       "      <td>MEDIUM.7H9:1_ETHANOL:0.2vol%_STRAIN.H37Rv:1_TE...</td>\n",
       "      <td>test_GSE1642_TRICLO100</td>\n",
       "      <td>ref_GSE1642_</td>\n",
       "      <td>GSE1642</td>\n",
       "      <td>GEO</td>\n",
       "      <td>GPL1396</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    contrast_name  \\\n",
       "345  GSM28131.ch1-vs-GSM28131.ch2   \n",
       "\n",
       "                                            annotation  \\\n",
       "345  MEDIUM.7H9:1_ETHANOL:0.2vol%_STRAIN.H37Rv:1_TE...   \n",
       "\n",
       "           Test description Reference description Experiment_id Data source  \\\n",
       "345  test_GSE1642_TRICLO100          ref_GSE1642_       GSE1642         GEO   \n",
       "\n",
       "    Platform  \n",
       "345  GPL1396  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rna_annot.iloc[[345]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
